Fixes SI-6260

odersky opened the pull request:
and also has a small cleanup for OverridingPairs. Review by @paulp @retronym

odersky committed:
Fixes SI-6260

Guards against bridge methods that clash with other methods. Two tests: The neg test is the original ticket. The run test tweaks things slightly so that the generated bridge method does not clash, and tests that the necessary unboxings are indeed performed at runtime.

odersky committed:
Cleanup of OverridingPairs

Added a default value to a hashmap so that we can work with it without Option acrobatics.

paulp commented on src/compiler/scala/tools/nsc/transform/OverridingPairs.scala:
```diff
@@ -104,8 +104,11 @@ abstract class OverridingPairs {
+     *  symbols that are not baseclasses map to -1.
      */
-    private val index = new mutable.HashMap[Symbol, Int]
+    private val index = new mutable.HashMap[Symbol, Int] {
```
Using withDefaultValue spares us an extra anonymous class.

paulp commented:
This looks solid to me.  The only thing I'd like to do is have the error message say something less low-level about why it happened and what can be done to avoid it; I'll look at that myself.  LGTM.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/464/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1171/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/464/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1171/

jsuereth commented on test/files/neg/t6260.check:
```diff
@@ -0,0 +1,13 @@
+t6260.scala:3: error: bridge generated for member method apply: (x$1: Box[X])Box[Y] in anonymous class $anonfun
```
If you didn't understand the compiler or Scala all that well, just learned about AnyVal classes and decide to try them.  Then you run into this error message, would you have any idea what you were dealing with or how to fix it?

Since we have the opportunity to offer a specific suggestion or just say "don't do this", can we alter the error message?

paulp replied:
That's what my comment says (and implies that I'll do it.)

scala-jenkins replied:
Time sync error.   You response appears first, but didn't show up when I
was making mine.

gkossakowski replied:
@paulp: Any estimates for time you need for it? I'm asking because I'd prefer putting this pull request on hold until we see error message improvement. I agree with you and @jsuereth that user should never be exposed error messages like this one. I'm afraid that once this pull request gets merged the error message problem wil get forgotten.

odersky committed:
Value classes: eliminated half-boxing

We now apply erasure of value classes everywhere. previously, erasure was disabled in the value class itself. This led to irregegularities and bugs. See test run/valueclasses-pavlov.scala	for something that led to a ClassCastException before.
